backpack-lint/src/main/java/net/skyscanner/backpack/lint/check/BpkComponentUsageDetector.kt

```diff
@@ -0,0 +1,108 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.skyscanner.backpack.lint.check
+
+import com.android.resources.ResourceFolderType
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Context
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Location
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.XmlContext
+import com.android.tools.lint.detector.api.XmlScanner
+import com.intellij.psi.PsiMethod
+import org.jetbrains.uast.UCallExpression
+import org.jetbrains.uast.UClass
+import org.w3c.dom.Element
+
+@Suppress("UnstableApiUsage")
+class BpkComponentUsageDetector : Detector(), SourceCodeScanner, XmlScanner {
+
+  companion object {
+    val ISSUE = Issue.create(
+      id = "BpkComponentUsage",
+      briefDescription = "A backpack version of this component is available",
+      explanation = "Use backpack components where available to improve consistency, accessibility and dark mode support.",
+      category = Category.CORRECTNESS,
+      severity = Severity.WARNING,
+      implementation = Implementation(
+        BpkComponentUsageDetector::class.java,
+        Scope.JAVA_AND_RESOURCE_FILES
+      )
+    )
+
+    private const val BPK_BUTTON = "net.skyscanner.backpack.BpkButton"
+
+    private val COMPONENTS = mapOf(
+      // BpkButton
+      "android.widget.Button" to BPK_BUTTON,
+      "Button" to BPK_BUTTON,
+      "androidx.appcompat.widget.AppCompatButton" to BPK_BUTTON,
+    )
+
+    private val APPLICABLE_TYPES = COMPONENTS.keys.toList()
+  }
+
+  private val classCache = mutableListOf<String>()
+
+  override fun getApplicableConstructorTypes(): List<String> = APPLICABLE_TYPES
+
+  override fun applicableSuperClasses(): List<String> = APPLICABLE_TYPES
+
+  override fun getApplicableElements(): Collection<String> = APPLICABLE_TYPES
+
+  override fun appliesTo(folderType: ResourceFolderType): Boolean = folderType == ResourceFolderType.LAYOUT
+
+  override fun visitClass(context: JavaContext, declaration: UClass) {
+    val superClass = declaration.javaPsi.superClass?.qualifiedName ?: return
+    val component = COMPONENTS[superClass]
+    if (component != null) {
+      if (declaration.name == null || classCache.contains(declaration.name)) {
+        // in some cases we might have two relevant super classes (e.g. when there's an AppCompat and default component) - only report issue once
```

Comment:
As `AppCompatButton` extends `Button` the `applicableSuperClasses` check gets triggered twice. This is the best way I could find to fix this - we _do_ want to check that if you're extending either `Button` or `ApCompatButton` we trigger the check, but there doesn't seem to be any hint about which one triggered, so we can compare it with the class that was actually extended. So keeping a memory of checks that already is the only thing I could think of to not report the error twice